OMEdit/OMEdit/OMEditGUI/Annotations/TextAnnotation.cpp

```diff
@@ -300,96 +300,88 @@ void TextAnnotation::drawTextAnnotaion(QPainter *painter)
-  // first we invert the painter since we have our coordinate system inverted.
-  painter->scale(1.0, -1.0);
-  painter->translate(0, dy);
+  // store the existing transformations
+  const QTransform painterTransform = painter->transform();
+  const qreal scaleX = painterTransform.m11();
+  const qreal scaleY = painterTransform.m22();
+  const qreal shearX = painterTransform.m21();
+  const qreal shearY = painterTransform.m12();
+  // set new transformation for the text based on rotation or scale.
+  qreal sx, sy;
+  if (painterTransform.type() == QTransform::TxRotate) {
+    // if no flip or XY flip
+    if ((shearX >= 0 && shearY >= 0) || (shearX < 0 && shearY < 0)) {
+      painter->setTransform(QTransform(painterTransform.m11(), 1.0 * (shearY < 0 ? -1.0 : 1.0), painterTransform.m13(),
+                                       1.0 * (shearX >= 0 ? -1.0 : 1.0), painterTransform.m22(), painterTransform.m23(),
+                                       painterTransform.m31(), painterTransform.m32(), painterTransform.m33()));
+      sx = shearX * (shearX < 0 ? -1.0 : 1.0);
+      sy = shearY * (shearY >= 0 ? -1.0 : 1.0);
+    } else { // if x or y flip
+      painter->setTransform(QTransform(painterTransform.m11(), 1.0 * (shearY < 0 ? 1.0 : -1.0), painterTransform.m13(),
+                                       1.0 * (shearX >= 0 ? -1.0 : 1.0), painterTransform.m22(), painterTransform.m23(),
+                                       painterTransform.m31(), painterTransform.m32(), painterTransform.m33()));
+      sx = shearX * (shearX < 0 ? 1.0 : -1.0);
+      sy = shearY * (shearY >= 0 ? -1.0 : 1.0);
+    }
+  } else {
+    painter->setTransform(QTransform(1.0, painterTransform.m12(), painterTransform.m13(),
+                                     painterTransform.m21(), 1.0, painterTransform.m23(),
+                                     painterTransform.m31(), painterTransform.m32(), painterTransform.m33()));
+    sx = scaleX;
+    sy = scaleY;
+  }
+  // map the existing bounding rect to new transformation
+  QRectF mappedBoundingRect = QRectF(boundingRect().x() * sx, boundingRect().y() * sy, boundingRect().width() * sx, boundingRect().height() * sy);
+  // map the existing bounding rect to new transformation but with positive width and height so that font metrics can work
+  QRectF absMappedBoundingRect = QRectF(boundingRect().x() * sx, boundingRect().y() * sy, qAbs(boundingRect().width() * sx), qAbs(boundingRect().height() * sy));
+  // normalize the text for drawing
   mTextString = StringHandler::removeFirstLastQuotes(mTextString);
   mTextString = StringHandler::unparse(QString("\"").append(mTextString).append("\""));
-  QFont font;
-  font = QFont(mFontName, mFontSize, StringHandler::getFontWeight(mTextStyles), StringHandler::getFontItalic(mTextStyles));
+  // Don't create new QFont instead get a font from painter and set the values on it and set it back.
+  QFont font = painter->font();
```

Comment:
This font handling code would be nice to have in a separate function as well. We let it configure the front and return it to the painter so it can set it on line 352